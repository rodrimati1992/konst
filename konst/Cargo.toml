[package]
name = "konst"
version = "0.3.5"
authors = ["rodrimati1992 <rodrimatt1985@gmail.com>"]
rust-version = "1.65.0"
edition = "2021"
license = "Zlib"
description = "Const equivalents of std functions, compile-time comparison, and parsing"
documentation = "https://docs.rs/konst/"
readme="../README.md"
keywords = ["no-std", "const_fn", "parsing"]
categories = ["no-std", "parsing"]
repository = "https://github.com/rodrimati1992/konst/"
include = [
    "Cargo.toml", 
    "src/**/*.rs", 
    "../README.md",
    "LICENSE-ZLIB.md", 
]


[dependencies.konst_proc_macros]
version = "=0.3.0"
path = "../konst_proc_macros"
optional = true

[dependencies.konst_kernel]
version = "=0.3.5"
path = "../konst_kernel"
features = ["rust_1_64", "__for_konst"]

[dev-dependencies.rand]
version = "0.8.4"
default-features = false
features = ["small_rng"]

# dev-dependencies can't be optional
[dependencies.trybuild]
version = "1.0"
optional = true

[dependencies.typewit]
version = "1.1"
features = ["rust_1_61"]

[dependencies.const_panic]
version = "0.2.6"
features = ["rust_1_64"]
default-features = false

[features]
default = ["cmp", "iter", "parsing_proc"]


# Enables extra checks for debug
debug = ["konst_kernel/debug"]
cmp = []
iter = ["konst_kernel/iter"]
parsing_proc = ["parsing", "konst_proc_macros"]
parsing = []
rust_latest_stable = []
mut_refs = ["rust_latest_stable", "konst_kernel/mut_refs"]
nightly_mut_refs = ["mut_refs", "konst_kernel/nightly_mut_refs"]

alloc = ["konst_kernel/alloc"]

# Enables ui tests, which are intended for the latest stable release.
# This is not enabled in CI, because it's way too sensitive to 
# changes in how rustc formats errors
__ui = ["trybuild", "rust_latest_stable"]
__cp_derive = ["const_panic/derive", "const_panic/non_basic"]

docsrs = ["konst_kernel/docsrs"]



[package.metadata.docs.rs]
features = ["docsrs", "alloc", "cmp", "parsing_proc", "rust_latest_stable", "nightly_mut_refs"]

