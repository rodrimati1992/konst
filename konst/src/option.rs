//! `const` equivalents of `Option` methods.

/// A const equivalent of [`Option::unwrap`]
///
/// # Example
///
/// ```rust
/// use konst::option::unwrap;
///
/// use std::num::NonZeroUsize;
///
/// const TEN: NonZeroUsize = unwrap!(NonZeroUsize::new(10));
///
/// assert_eq!(TEN.get(), 10);
/// ```
#[doc(inline)]
pub use konst_kernel::opt_unwrap as unwrap;

/// A const equivalent of [`Option::unwrap_or`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[u32] = &[
///     option::unwrap_or!(Some(3), 10000),
///     option::unwrap_or!(None, 5),
/// ];
///
/// assert_eq!(ARR, &[3, 5]);
///
/// ```
///
#[doc(inline)]
pub use konst_kernel::opt_unwrap_or as unwrap_or;

/// A const equivalent of [`Option::unwrap_or_else`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[u32] = &[
///     // You can use a closure-like syntax to run code when the Option argument is None.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::unwrap_or_else!(Some(3), || loop{}),
///     option::unwrap_or_else!(None, || 5),
///
///     // You can also pass functions
///     option::unwrap_or_else!(Some(8), thirteen),
///     option::unwrap_or_else!(None, thirteen),
/// ];
///
/// assert_eq!(ARR, &[3, 5, 8, 13]);
///
/// const fn thirteen() -> u32 {
///     13
/// }
/// ```
///
#[doc(inline)]
pub use konst_kernel::opt_unwrap_or_else as unwrap_or_else;

/// A const equivalent of [`Option::ok_or`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Result<u32, u32>] = &[
///     option::ok_or!(Some(3), 10000),
///     option::ok_or!(None, 5),
/// ];
///
/// assert_eq!(ARR, &[Ok(3), Err(5)]);
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_ok_or as ok_or;

/// A const equivalent of [`Option::ok_or_else`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Result<u32, u32>] = &[
///     // You can use a closure-like syntax to run code when the Option argument is None.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::ok_or_else!(Some(3), || loop{}),
///     option::ok_or_else!(None, || 5),
///
///     // You can also pass functions
///     option::ok_or_else!(Some(8), thirteen),
///     option::ok_or_else!(None, thirteen),
/// ];
///
/// assert_eq!(ARR, &[Ok(3), Err(5), Ok(8), Err(13)]);
///
/// const fn thirteen() -> u32 {
///     13
/// }
/// ```
#[doc(inline)]
pub use konst_kernel::opt_ok_or_else as ok_or_else;

/// A const equivalent of [`Option::map`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     // You can use a closure-like syntax to pass code that maps the Some variant.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::map!(Some(3), |x| x * 3),
///     option::map!(None::<u32>, |_| loop{}),
///
///     // You can also pass functions
///     option::map!(Some(8), double),
///     option::map!(None::<u32>, double),
/// ];
///
/// assert_eq!(ARR, &[Some(9), None, Some(16), None]);
///
/// const fn double(x: u32) -> u32 {
///     x * 2
/// }
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_map as map;

/// A const equivalent of [`Option::and_then`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     // You can use a closure-like syntax to pass code that uses the value in the Some variant.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::and_then!(Some(3), |x| Some(x * 3)),
///     option::and_then!(Some(3), |_| None),
///     option::and_then!(None::<u32>, |_| loop{}),
///
///     // You can also pass functions
///     option::and_then!(Some(23), checked_sub),
///     option::and_then!(Some(9), checked_sub),
///     option::and_then!(None::<u32>, checked_sub),
/// ];
///
/// assert_eq!(ARR, &[Some(9), None, None, Some(13), None, None]);
///
/// const fn checked_sub(x: u32) -> Option<u32> {
/// # /*
///     x.checked_sub(10)
/// # */
/// #   let (ret, overflowed) = x.overflowing_sub(10);
/// #   if overflowed { None } else { Some(ret) }
/// }
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_and_then as and_then;

/// A const equivalent of [`Option::or_else`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     // You can use a closure-like syntax to pass code that runs on None.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::or_else!(Some(3), || loop{}),
///     option::or_else!(None::<u32>, || Some(5)),
///
///     // You can also pass functions
///     option::or_else!(Some(8), thirteen),
///     option::or_else!(None::<u32>, thirteen),
/// ];
///
/// assert_eq!(ARR, &[Some(3), Some(5), Some(8), Some(13)]);
///
/// const fn thirteen() -> Option<u32> {
///     Some(13)
/// }
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_or_else as or_else;

/// A const equivalent of [`Option::flatten`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     option::flatten!(Some(Some(8))),
///     option::flatten!(None),
/// ];
///
/// assert_eq!(ARR, &[Some(8), None]);
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_flatten as flatten;

/// A const equivalent of [`Option::filter`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     // You can use a closure-like syntax to pass code that filters the Some variant.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::filter!(Some(0), |&x| x == 0),
///     option::filter!(Some(1), |x| *x == 0),
///     option::filter!(None, |_| loop{}),
///
///     // You can also pass functions
///     option::filter!(Some(3), is_odd),
///     option::filter!(Some(4), is_odd),
///     option::filter!(None, is_odd),
/// ];
///
/// assert_eq!(ARR, &[Some(0), None, None, Some(3), None, None]);
///
/// const fn is_odd(x: &u32) -> bool {
///     *x % 2 == 1
/// }
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_filter as filter;

/// A const equivalent of the [`Option::copied`] method.
///
/// # Example
///
/// ```rust
/// use konst::option;
///
/// const fn get_last(slice: &[u64]) -> Option<u64> {
///     option::copied(slice.last())
/// }
///
/// assert_eq!(get_last(&[]), None);
/// assert_eq!(get_last(&[16]), Some(16));
/// assert_eq!(get_last(&[3, 5, 8, 13]), Some(13));
///
///
/// ```
pub const fn copied<T: Copy>(opt: Option<&T>) -> Option<T> {
    match opt {
        Some(x) => Some(*x),
        None => None,
    }
}

declare_generic_const! {
    /// Usable to do `[None::<T>; LEN]` when `T` is non-`Copy`.
    ///
    /// As of Rust 1.65.0, `[None::<T>; LEN]` is not valid for non-`Copy` types,
    /// but `[CONST; LEN]` does work, like in the example below.
    ///
    /// # Example
    ///
    /// ```rust
    /// use konst::option::NONE;
    ///
    /// use std::mem::{self, MaybeUninit};
    ///
    /// const TEN: [Option<String>; 10] = [NONE::V; 10];
    ///
    /// let ten = [NONE::<String>::V; 10];
    ///
    /// // the `vec` macro only needs `Option<String>` to be clonable, not Copy.
    /// assert_eq!(vec![None::<String>; 10], TEN);
    ///
    /// assert_eq!(vec![None::<String>; 10], ten);
    ///
    for[T]
    pub const NONE[T]: Option<T> = None;
}
