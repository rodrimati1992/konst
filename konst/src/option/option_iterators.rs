use crate::iter::{ConstIntoIter, IntoIterWrapper, IsIteratorKind, IsStdKind};

use core::mem::ManuallyDrop;



impl<'a, T> ConstIntoIter for &'a Option<T> {
    type Kind = IsStdKind;
    type IntoIter = Iter<'a, T>;
    type Item = &'a T;
}

impl<'a, T> ConstIntoIter for &&'a Option<T> {
    type Kind = IsStdKind;
    type IntoIter = Iter<'a, T>;
    type Item = &'a T;
}

impl<'a, T> IntoIterWrapper<&'a Option<T>, IsStdKind> {
    /// Converts `&'a Option<T>` into an iterator
    pub const fn const_into_iter(self) -> Iter<'a, T> {
        Iter {
            opt: ManuallyDrop::into_inner(self.iter).as_ref(),
        }
    }
}
impl<'a, T> IntoIterWrapper<&&'a Option<T>, IsStdKind> {
    /// Converts `&&'a Option<T>` into an iterator
    pub const fn const_into_iter(self) -> Iter<'a, T> {
        Iter {
            opt: (*ManuallyDrop::into_inner(self.iter)).as_ref(),
        }
    }
}


/// Const equivalent of [`Option::iter`]
/// 
/// # Example
/// 
/// ```rust
/// use konst::option;
/// 
/// let mut fwd = option::iter(&Some(5));
/// assert_eq!(fwd.next(), Some(&5));
/// assert_eq!(fwd.next(), None);
/// 
/// let mut rev = option::iter(&Some(8)).rev();
/// assert_eq!(rev.next(), Some(&8));
/// assert_eq!(rev.next(), None);
/// ```
pub const fn iter<T>(opt: &Option<T>) -> Iter<'_, T> {
    Iter {
        opt: opt.as_ref()
    }
}

macro_rules! iter_shared {
    (is_forward = $is_forward:ident) => {
        crate::iterator_shared! {
            is_forward = $is_forward,
            item = &'a T,
            iter_forward = Iter<'a, T>,
            iter_reversed = IterRev<'a, T>,
            next(self) {
                self.opt.take()
            },
            next_back {
                self.opt.take()
            },
            fields = {opt},
        }
    };
}

/// Const equivalent of [`core::option::Iter`]
pub struct Iter<'a, T> {
    opt: Option<&'a T>,
}
impl<'a, T> ConstIntoIter for Iter<'a, T> {
    type Kind = IsIteratorKind;
    type IntoIter = Self;
    type Item = &'a T;
}

/// Const equivalent of `core::iter::Rev<core::option::Iter<T>>`
pub struct IterRev<'a, T> {
    opt: Option<&'a T>,
}
impl<'a, T> ConstIntoIter for IterRev<'a, T> {
    type Kind = IsIteratorKind;
    type IntoIter = Self;
    type Item = &'a T;
}

impl<'a, T> Iter<'a, T> {
    iter_shared! {is_forward = true}
}

impl<'a, T> IterRev<'a, T> {
    iter_shared! {is_forward = false}
}

////////////////////////////////////////////////////////////////////////////////

impl<'a, T> ConstIntoIter for &'a mut Option<T> {
    type Kind = IsStdKind;
    type IntoIter = IterMut<'a, T>;
    type Item = &'a mut T;
}

impl<'a, T> ConstIntoIter for &'a mut &mut Option<T> {
    type Kind = IsStdKind;
    type IntoIter = IterMut<'a, T>;
    type Item = &'a mut T;
}

impl<'a, T> IntoIterWrapper<&'a mut Option<T>, IsStdKind> {
    /// Converts `&'a mut Option<T>` into an iterator
    pub const fn const_into_iter(self) -> IterMut<'a, T> {
        IterMut {
            opt: ManuallyDrop::into_inner(self.iter).as_mut(),
        }
    }
}
impl<'a, T> IntoIterWrapper<&'a mut &mut Option<T>, IsStdKind> {
    /// Converts `&'a mut &mut Option<T>` into an iterator
    pub const fn const_into_iter(self) -> IterMut<'a, T> {
        IterMut {
            opt: ManuallyDrop::into_inner(self.iter).as_mut(),
        }
    }
}


/// Const equivalent of [`Option::iter_mut`]
/// 
/// # Example
/// 
/// ```rust
/// use konst::option;
/// 
/// {
///     let mut opt = Some(13);
///     let mut fwd = option::iter_mut(&mut opt);
///     assert_eq!(fwd.next(), Some(&mut 13));
///     assert_eq!(fwd.next(), None);
/// }
/// {
///     let mut opt = Some(21);
///     let mut rev = option::iter_mut(&mut opt).rev();
///     assert_eq!(rev.next(), Some(&mut 21));
///     assert_eq!(rev.next(), None);
/// }
/// ```
pub const fn iter_mut<T>(opt: &mut Option<T>) -> IterMut<'_, T> {
    IterMut {
        opt: opt.as_mut()
    }
}

macro_rules! iter_mut_shared {
    (is_forward = $is_forward:ident) => {
        iterator_shared! {
            is_forward = $is_forward,
            is_copy = false,
            item = &'a mut T,
            iter_forward = IterMut<'a, T>,
            iter_reversed = IterMutRev<'a, T>,
            next(self) {
                self.opt.take()
            },
            next_back {
                self.opt.take()
            },
            fields = {opt},
        }
    };
}

/// Const equivalent of [`core::option::IterMut`]
pub struct IterMut<'a, T> {
    opt: Option<&'a mut T>,
}
impl<'a, T> ConstIntoIter for IterMut<'a, T> {
    type Kind = IsIteratorKind;
    type IntoIter = Self;
    type Item = &'a mut T;
}

/// Const equivalent of `core::iter::Rev<core::option::IterMut<T>>`
pub struct IterMutRev<'a, T> {
    opt: Option<&'a mut T>,
}
impl<'a, T> ConstIntoIter for IterMutRev<'a, T> {
    type Kind = IsIteratorKind;
    type IntoIter = Self;
    type Item = &'a mut T;
}

impl<'a, T> IterMut<'a, T> {
    iter_mut_shared! {is_forward = true}
}

impl<'a, T> IterMutRev<'a, T> {
    iter_mut_shared! {is_forward = false}
}

////////////////////////////////////////////////////////////////////////////////

impl<T> ConstIntoIter for Option<T> {
    type Kind = IsStdKind;
    type IntoIter = IntoIter<T>;
    type Item = T;
}

impl<T> IntoIterWrapper<Option<T>, IsStdKind> {
    /// Converts `Option<T>` into an iterator
    pub const fn const_into_iter(self) -> IntoIter<T> {
        IntoIter {
            opt: ManuallyDrop::into_inner(self.iter),
        }
    }
}

/// Const equivalent of [`Option::into_iter`]
/// 
/// # Example
/// 
/// ```rust
/// use konst::option;
/// 
/// {
///     let mut opt = Some(13);
///     let mut fwd = option::into_iter(opt);
///     assert_eq!(fwd.next(), Some(13));
///     assert_eq!(fwd.next(), None);
/// }
/// {
///     let mut opt = Some(21);
///     let mut rev = option::into_iter(opt).rev();
///     assert_eq!(rev.next(), Some(21));
///     assert_eq!(rev.next(), None);
/// }
/// ```
pub const fn into_iter<T>(opt: Option<T>) -> IntoIter<T> {
    IntoIter { opt }
}

macro_rules! into_iter_shared {
    (is_forward = $is_forward:ident) => {
        iterator_shared! {
            is_forward = $is_forward,
            is_copy = false,
            is_drop = true,
            item = T,
            iter_forward = IntoIter<T>,
            iter_reversed = IntoIterRev<T>,
            next(self) {
                self.opt.take()
            },
            next_back {
                self.opt.take()
            },
            fields = {opt},
        }
    };
}

/// Const equivalent of [`core::option::IntoIter`]
pub struct IntoIter<T> {
    opt: Option<T>,
}
impl<T> ConstIntoIter for IntoIter<T> {
    type Kind = IsIteratorKind;
    type IntoIter = Self;
    type Item = T;
}

/// Const equivalent of `core::iter::Rev<core::option::IntoIter<T>>`
pub struct IntoIterRev<T> {
    opt: Option<T>,
}
impl<T> ConstIntoIter for IntoIterRev<T> {
    type Kind = IsIteratorKind;
    type IntoIter = Self;
    type Item = T;
}

impl<T> IntoIter<T> {
    into_iter_shared! {is_forward = true}
}

impl<T> IntoIterRev<T> {
    into_iter_shared! {is_forward = false}
}

