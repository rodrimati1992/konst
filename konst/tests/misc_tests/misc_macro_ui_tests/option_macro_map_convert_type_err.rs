use konst::option;

const _: Result<(), u16> = option::ok_or!((), 0u16);
const _: () = option::ok_or!(None::<u8>, 0u16);

const _: Result<(), u16> = option::ok_or_else!((), || 0u16);
const _: () = option::ok_or_else!(None::<u8>, || 0u16);

const _: Option<u8> = option::map!((), |_| 0u8);
const _: Option<u8> = option::map!(None::<()>, |__ @ 0u8..| 0u8);
const _: () = option::map!(None::<u8>, |_| ());

const _: Option<u16> = option::and_then!((), |_| None::<u16>);
const _: Option<u16> = option::and_then!(None::<()>, |__ @ 0u8..| None::<u16>);
const _: () = option::and_then!(None::<u8>, |__ @ 0u8..| None::<u16>);

const _: Option<u16> = option::or_else!((), || None::<u16>);
const _: Option<u16> = option::or_else!(None::<()>, || None::<u16>);
const _: () = option::or_else!(None::<u8>, || None::<u16>);


fn main() {}
