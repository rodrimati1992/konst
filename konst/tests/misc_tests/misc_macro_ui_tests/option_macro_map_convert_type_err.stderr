error[E0308]: mismatched types
 --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:3:43
  |
3 | const _: Result<(), u16> = option::ok_or!((), 0u16);
  |                            ---------------^^-------
  |                            |              |
  |                            |              expected `Option<_>`, found `()`
  |                            arguments to this function are incorrect
  |
  = note:   expected enum `Option<_>`
          found unit type `()`
note: function defined here
 --> src/option.rs
  |
  | pub const fn __opt<T>(opt: Option<T>) -> Option<T> {
  |              ^^^^^
help: try wrapping the expression in `Some`
  |
3 | const _: Result<(), u16> = option::ok_or!(Some(()), 0u16);
  |                                           +++++  +

error[E0308]: mismatched types
 --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:4:15
  |
4 | const _: () = option::ok_or!(None::<u8>, 0u16);
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `Result<u8, u16>`
  |
  = note: expected unit type `()`
                  found enum `Result<u8, u16>`
  = note: this error originates in the macro `option::ok_or` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
 --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:6:48
  |
6 | const _: Result<(), u16> = option::ok_or_else!((), || 0u16);
  |                            --------------------^^----------
  |                            |                   |
  |                            |                   expected `Option<_>`, found `()`
  |                            arguments to this function are incorrect
  |
  = note:   expected enum `Option<_>`
          found unit type `()`
note: function defined here
 --> src/option.rs
  |
  | pub const fn __opt<T>(opt: Option<T>) -> Option<T> {
  |              ^^^^^
help: try wrapping the expression in `Some`
  |
6 | const _: Result<(), u16> = option::ok_or_else!(Some(()), || 0u16);
  |                                                +++++  +

error[E0308]: mismatched types
 --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:7:15
  |
7 | const _: () = option::ok_or_else!(None::<u8>, || 0u16);
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `Result<u8, _>`
  |
  = note: expected unit type `()`
                  found enum `Result<u8, _>`
  = note: this error originates in the macro `option::ok_or_else` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
 --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:7:15
  |
7 | const _: () = option::ok_or_else!(None::<u8>, || 0u16);
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `Result<_, u16>`
  |
  = note: expected unit type `()`
                  found enum `Result<_, u16>`
  = note: this error originates in the macro `option::ok_or_else` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
 --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:9:36
  |
9 | const _: Option<u8> = option::map!((), |_| 0u8);
  |                       -------------^^----------
  |                       |            |
  |                       |            expected `Option<_>`, found `()`
  |                       arguments to this function are incorrect
  |
  = note:   expected enum `Option<_>`
          found unit type `()`
note: function defined here
 --> src/option.rs
  |
  | pub const fn __opt<T>(opt: Option<T>) -> Option<T> {
  |              ^^^^^
help: try wrapping the expression in `Some`
  |
9 | const _: Option<u8> = option::map!(Some(()), |_| 0u8);
  |                                    +++++  +

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:10:54
   |
10 | const _: Option<u8> = option::map!(None::<()>, |__ @ 0u8..| 0u8);
   |                       -------------------------------^^^--------
   |                       |                              |
   |                       |                              expected `()`, found `u8`
   |                       this expression has type `()`

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:11:15
   |
11 | const _: () = option::map!(None::<u8>, |_| ());
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `Option<()>`
   |
   = note: expected unit type `()`
                   found enum `Option<()>`
   = note: this error originates in the macro `option::map` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:11:15
   |
11 | const _: () = option::map!(None::<u8>, |_| ());
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `Option<_>`
   |
   = note: expected unit type `()`
                   found enum `Option<_>`
   = note: this error originates in the macro `option::map` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:13:42
   |
13 | const _: Option<u16> = option::and_then!((), |_| None::<u16>);
   |                        ------------------^^------------------
   |                        |                 |
   |                        |                 expected `Option<_>`, found `()`
   |                        arguments to this function are incorrect
   |
   = note:   expected enum `Option<_>`
           found unit type `()`
note: function defined here
  --> src/option.rs
   |
   | pub const fn __opt<T>(opt: Option<T>) -> Option<T> {
   |              ^^^^^
help: try wrapping the expression in `Some`
   |
13 | const _: Option<u16> = option::and_then!(Some(()), |_| None::<u16>);
   |                                          +++++  +

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:14:60
   |
14 | const _: Option<u16> = option::and_then!(None::<()>, |__ @ 0u8..| None::<u16>);
   |                        ------------------------------------^^^----------------
   |                        |                                   |
   |                        |                                   expected `()`, found `u8`
   |                        this expression has type `()`

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:15:15
   |
15 | const _: () = option::and_then!(None::<u8>, |__ @ 0u8..| None::<u16>);
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `Option<u16>`
   |
   = note: expected unit type `()`
                   found enum `Option<u16>`
   = note: this error originates in the macro `option::and_then` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:15:15
   |
15 | const _: () = option::and_then!(None::<u8>, |__ @ 0u8..| None::<u16>);
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `Option<_>`
   |
   = note: expected unit type `()`
                   found enum `Option<_>`
   = note: this error originates in the macro `option::and_then` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:17:41
   |
17 | const _: Option<u16> = option::or_else!((), || None::<u16>);
   |                        -----------------^^-----------------
   |                        |                |
   |                        |                expected `Option<_>`, found `()`
   |                        arguments to this function are incorrect
   |
   = note:   expected enum `Option<_>`
           found unit type `()`
note: function defined here
  --> src/option.rs
   |
   | pub const fn __opt<T>(opt: Option<T>) -> Option<T> {
   |              ^^^^^
help: try wrapping the expression in `Some`
   |
17 | const _: Option<u16> = option::or_else!(Some(()), || None::<u16>);
   |                                         +++++  +

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:18:24
   |
18 | const _: Option<u16> = option::or_else!(None::<()>, || None::<u16>);
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<u16>`, found `Option<()>`
   |
   = note: expected enum `Option<u16>`
              found enum `Option<()>`
   = note: this error originates in the macro `option::or_else` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:18:56
   |
18 | const _: Option<u16> = option::or_else!(None::<()>, || None::<u16>);
   |                        --------------------------------^^^^^^^^^^^-
   |                        |                               |
   |                        |                               expected `Option<()>`, found `Option<u16>`
   |                        arguments to this function are incorrect
   |
   = note: expected enum `Option<()>`
              found enum `Option<u16>`
note: function defined here
  --> src/__utils.rs
   |
   | pub const fn __overwrite<T>(mr: &mut T, val: T) {
   |              ^^^^^^^^^^^

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:18:24
   |
18 | const _: Option<u16> = option::or_else!(None::<()>, || None::<u16>);
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<u16>`, found `Option<()>`
   |
   = note: expected enum `Option<u16>`
              found enum `Option<()>`
   = note: this error originates in the macro `option::or_else` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:19:47
   |
19 | const _: () = option::or_else!(None::<u8>, || None::<u16>);
   |               --------------------------------^^^^^^^^^^^-
   |               |                               |
   |               |                               expected `Option<u8>`, found `Option<u16>`
   |               arguments to this function are incorrect
   |
   = note: expected enum `Option<u8>`
              found enum `Option<u16>`
note: function defined here
  --> src/__utils.rs
   |
   | pub const fn __overwrite<T>(mr: &mut T, val: T) {
   |              ^^^^^^^^^^^

error[E0308]: mismatched types
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_map_convert_type_err.rs:19:15
   |
19 | const _: () = option::or_else!(None::<u8>, || None::<u16>);
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `Option<u8>`
   |
   = note: expected unit type `()`
                   found enum `Option<u8>`
   = note: this error originates in the macro `option::or_else` (in Nightly builds, run with -Z macro-backtrace for more info)
